### DISCUSSION TITLE
Modifying a large GeoTIFF dataset while utilizing open_mfdataset()

### DISCUSSION THREAD

**Discussion by Doomski99** [corteva/rioxarray#496] [Q&A] [answered]
Hello everyone,

I hope everyone is having a wonderful day. I have a large (25K) dataset of GeoTIFFs each consisting of 206 bands. I need to access each image and remove a couple of bands. I wrote first a simple for loop for the task but my ram quickly got maxed and the operation was very slow.  This is my for loop implementation:
```
for root, dirs, files in os.walk(filepath):
    for file in files:
        da = rioxr.open_rasterio(os.path.join(filepath, file))

        da = da.drop(indices_toremove, dim='band')

        da.attrs['long_name'] = descr

        da.rio.to_raster(os.path.join(filepath2, file))

        del da
```
EDIT: I fixed the ram issue by wrapping with `with` : 
```
for root, dirs, files in os.walk(filepath):
    for file in files:
        with rioxr.open_rasterio(os.path.join(filepath, file)) as da:

            da = da.drop_sel(band = indices_toremove)

            da.attrs['long_name'] = descr

            da.rio.to_raster(os.path.join(filepath2, file))
```

So I decided to utilize **Dask** since **xarray** supports it (still new to Dask, sorry if what I wrote doesn't make sense). But I ran into 2 problems and I need your guidance if possible; I want to know if my way of programming is right or wrong.

The first problem I've ran into is, while testing a single GeoTIFF with `open_mfdataset()`; when I try to write it with `rio.to_raster()`, I get the `TooManyDimensions` error, why is that? The docs state it only supports 2D data, but when I used `open_rasterio()` to open the tif, `rio.to_raster()` works properly. 

Now that aside, if I'm using `open_mfdataset()` to open many files, how can I use `rio.to_raster()` without stating the file names since all of them were loaded at once into dask? The following is the script I've wrote, maybe it will clarify the way my problem:
```
filepath = 'data\\raw\\*.tif'
da = xr.open_mfdataset(filepath, parallel = True,engine = 'rasterio')
da = da.drop_sel(band = indices_toremove)
da.rio.to_raster(??????)
```
And here's a representation of the data when loaded:
![image](https://user-images.githubusercontent.com/45564681/161386906-6d37e133-7b15-4f63-8556-bdc446754e55.png)

Can anyone help me?
Thank you!

**Comment by snowman2** [comment, marked as answer]
Since you are opening each individual file, modifying it, and writing to disk, I believe dask/open_mfdataset are not the tools for you. Dask is meant for reading a single large file or a collection of files that behave like a single large file (example: Zarr). open_mfdataset is for opening a collection of files that are meant to be joined together by one of the dimensions, such as "time".

I think your use case would be better served with multiprocessing/multithreading where you create a function that takes an input file path and makes the modifications and then map that function to a list of paths.

**Reply by Doomski99** [reply]
I see, thank you!
